rust/otel-arrow-rust/src/otlp/traces/span_links_store.rs

```diff
@@ -0,0 +1,624 @@
+use std::collections::HashMap;
+
+#[derive(Default)]
+pub struct SpanLinksStore {
+    links_by_id: HashMap<u16, Vec<Link>>,
+}
+
+impl SpanLinksStore {
+    pub(crate) fn link_by_id(&self, id: u16) -> Option<&Vec<Link>> {
+        self.links_by_id.get(&id)
+    }
+}
+
+/// Creates a `SpanLinksStore` from an Arrow RecordBatch containing span link data.
+///
+/// This function processes a RecordBatch where each row represents a span link, extracting:
+/// - Parent span ID (decoded using delta encoding)
+/// - Linked span's trace ID and span ID  
+/// - Trace state
+/// - Dropped attributes count
+/// - Any associated attributes (if an Attribute32Store is provided)
+///
+/// The links are grouped by their parent span ID in the resulting store for efficient lookup.
+///
+/// # Parameters
+/// - `rb`: The RecordBatch containing span link data with expected schema:
+///   - ID (UInt32): Optional attribute ID
+///   - TRACE_ID (FixedSizeBinary[16]): Trace ID of linked span
+///   - PARENT_ID (UInt16): Delta-encoded parent span ID  
+///   - SPAN_ID (FixedSizeBinary[8]): Span ID of linked span
+///   - TRACE_STATE (String): Trace state string
+///   - DROPPED_ATTRIBUTES_COUNT (UInt32): Count of dropped attributes
+/// - `attr_store`: Optional mutable reference to an Attribute32Store for looking up attributes
+///
+/// # Returns
+/// - `Result<SpanLinksStore>`: Contains all span links grouped by parent span ID
+///
+/// # Errors
+/// Returns an error if:
+/// - Required columns are missing or have wrong data types
+/// - Delta decoding fails
+pub fn span_links_store_from_record_batch(
+    rb: &RecordBatch,
+    attr_store: Option<&mut Attribute32Store>,
+) -> error::Result<SpanLinksStore> {
+    let mut store = SpanLinksStore::default();
+    let iterator = SpanLinkIterator::try_new(rb, attr_store)?;
+
+    for batch in iterator {
+        let link = batch?;
+        store
+            .links_by_id
+            .entry(link.parent_id)
+            .or_default()
+            .push(link.link);
+    }
+    Ok(store)
+}
+
+struct SpanLinkIterator<'a> {
+    id: Option<&'a UInt32Array>,
+    trace_id: Option<FixedSizeBinaryArrayAccessor<'a>>,
+    parent_id: Option<&'a UInt16Array>,
+    span_id: Option<FixedSizeBinaryArrayAccessor<'a>>,
+    trace_state: Option<StringArrayAccessor<'a>>,
+    dropped_attributes_count: Option<UInt32ArrayAccessor<'a>>,
+    parent_id_decoder: DeltaDecoder<Vec<u8>, u16>,
+    attr_store: Option<&'a mut Attribute32Store>,
+    offset: usize,
+    num_rows: usize,
+}
+
+impl Iterator for SpanLinkIterator<'_> {
+    type Item = error::Result<SpanLink>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.num_rows {
+            return None;
+        }
+        let id = self.id.value_at(self.offset);
+        let parent_id = self.parent_id.value_at_or_default(self.offset);
+
+        // todo(v0y4g3r): maybe impl value_ref_at to avoid heap alloc.
```

Comment:
Side-note: It would be nice, in my opinion, if we could modify the Prost-generated code to use `[u8; 16]` as the type of a TraceID (and `[u8; 8]` for SpanIDs) though I recognize how this complicates error handling.